### ISSUE TITLE
API change: Add MMF status to FetchMatches

### ISSUE THREAD

**Issue by Laremere** [googleforgames/open-match#1017]
## Overview

Add a field to FetchMatchesResponse to indicate Match Function success/failure.  Never return which fails an entire FetchMatches call which stem from an individual mmf failing.  Change the contract so that any matches which had been streamed before the MMF returned an error are still sent to the synchronizer instead of failing the whole call immediately.

## Motivation

A match making function which has returned matches but then fails is not necessarily an indication that a whole FetchMatches call should fail.  A single FetchMatches may have a large number of profiles all succeeding.  Failing everything on a single mmf failure may block the whole system when a single unimportant mmfs pod fails.

An mmf may have too many tickets to finish processing before the proposal window closes.  In that case, it's very beneficial for any matches it had returned to properly be sent to the synchronizer and returned as a result (if passing the evaluator), so that the ticket pool continues to drain.

With the current mechanism, it's not possible to easily tell the root cause of a profile failing.

With optimistic stream from the mmf to the synchronizer before all mmfs have finished (which will increase overall system speed and scale), the synchronizer may already be marking tickets in matches as pending a match.  Failing at this point due to another MMF failing unnecessarily causes these tickets to be stuck waiting for the pending match time out to expire.

An alternative is to make the FetchMatches call only operate on a single profile (removing the repeated), though that still has some of the streaming problems.  Though they're less severe as they're specifically limited to the erroring mmf.

## Impact

Callers of FetchMatches will need to check the type of result returned.  They may ignore the mmf status, but they will likely want to additionally log this status.

## Change Proto

Change from:
```proto
message FetchMatchesResponse {
  // A Match generated by the user-defined MMF with the specified MatchProfiles.
  // A valid Match response will contain at least one ticket.
  Match match = 1;
}
```

To:
```proto
message FetchMatchesResponse {
  oneof response {
    // A Match generated by the user-defined MMF with the specified MatchProfiles.
    // A valid Match response will contain at least one ticket.
    Match match = 1;
    MmfResult mmf_result = 2;
  }
}

message MmfResult {
  string profile_name = 1;
  google.rpc.Status status = 2;
}
```
MmfResult would always be sent for each profile run.  In the future, we could expand this to allowing other information to be returned to the director.

Note that adding a oneof isn't backwards compatible, but I think it's better to have it and we're not 1.0 yet.

**Comment by Laremere** [comment]
I am also finding that it'd be good to return an evaluation status.  There's several cases in the synchronizer where it is reasonable for some matches to be returned, but also for an error to have occurred (eg, evaluator returns half the matches before timing out.)  In these cases, I'm currently having the backend return no errors.  The errors are being logged, but someone debugging may not notice the logs.

**Comment by calebatwd** [comment]
I like the idea of removing the bulk proposals and just moving to an individual profile run request

**Comment by Laremere** [comment]
I asked about this to the grpc team, and the response was the errors might shortcut messages being sent back.  Given that we want to do our best to process all matches we have and return them to the director, this doesn't seem acceptable.

They did mention the option of using trailing metadata.  It looks like a possibly nice approach, though I have two concerns:
1. Clients may not realize they need to handle it as it's not obviously part of the RPC.
2. Documentation says access to metadata is implementation dependent, which isn't great.

**Comment by Laremere** [comment]
```protobuf
message FetchMatchesResponse {
  // Fetch matches will return a stream of responses containing matches, followed by a
  // a single FetchSummary before closing the response stream.
  oneof response {
    // A Match generated by the user-defined MMF with the specified MatchProfiles.
    // A valid Match response will contain at least one ticket.
    Match match = 1; 

    // Details of the call.  Don't ingore this field!  See warning on FetchMatches
    // for more info.
    FetchSummary fetch_summary = 2;
  }
}

message FetchSummary {
  // Response status from running the matchfunction.  Will be present and OK in the event
  // of no errors.
  google.rpc.Status mmf_status = 1;

  // Response status from running the evaluator.  Will be present and OK in the event
  // of no errors.
  google.rpc.Status evaluator_status = 2;

  // Number of matches proposed by the match function.
  int64 proposals_generated = 3;

  // Number of matches which the evaluator accepted.
  int64 matches_accepted = 4;
}
```

```protobuf
service BackendService {
  // FetchMatches triggers a MatchFunction with the specified MatchProfile, which
  // returns a set of match proposals.  Those proposals are sent to the evaluator to
  // deduplicate tickets in matches.  Then matches are streamed back to the fetch
  // matches call.  The stream ends with the FetchSummary.
  //
  // WARNING: FetchMatches DOES NOT return an error if the match function or evaluator
  // return an error.  This is done so that if they return partial results, those results can be
  // processed.  Make sure to inspect the FetchSummary's status for both the match function
  // and evaluator to log and otherwise handle errors.
  rpc FetchMatches(FetchMatchesRequest) returns (stream FetchMatchesResponse) {
```

**Comment by sawagh** [comment]
Can you summarize how you expect errors to be handled? 

Given that streaming message ordering is not guaranteed (or is it?) - Could you receive a FetchSummary before receiving the last valid match? Should the check be just for FetchSummary and if any status fails, stop - or should you wait for the matches received to be also equal to the matches_accepted?

**Comment by Laremere** [comment]
Streaming order of messages is guaranteed.  You handle errors how they would be normally handled.

Grpc errors passed may bypass the message queue and that's the problem.

**Comment by sawagh** [comment]
Is the matches_accepted and proposals_generated just FYI then?

**Comment by Laremere** [comment]
Yeah, it's easy info to add.

**Comment by Laremere** [comment]
Actually, I think I want to add a third error status:
```protobuf
  // Response status from internal Open Match system.  Will contain any errors
  // not originating from the mmf or evaluator.
  google.rpc.Status system_status = 3;
```

**Comment by calebatwd** [comment]
I understand how an mmf might generate an error (like if there's something not caught and open match is going to indicate some quantity success + "run cut short by pipe closure" [or something]). But the evaluator runs across fetchMatches-requests. What's the intention of the evaluator status?

**Comment by calebatwd** [comment]
> Actually, I think I want to add a third error status:
> 
> ```protobuf
>   // Response status from internal Open Match system.  Will contain any errors
>   // not originating from the mmf or evaluator.
>   google.rpc.Status system_status = 3;
> ```

Would the system_status be on FetchMatchResponse (the parent)? Makes sense as a typical rpc response. :thumbs-up:

**Comment by calebatwd** [comment]
> Is the matches_accepted and proposals_generated just FYI then?

They're simple and top-level and I weirdly really like them. Feels like the kind of info I would expect on a "run summary" object. Another good outside-looking-in summary might be the run duration. We don't want to go to far down the road of blending our job-stats and our application metrics. Curious to hear y'alls thoughts on what the bar might be

**Comment by calebatwd** [comment]
Looks like a good improvement!

**Comment by Laremere** [comment]
> Would the system_status be on FetchMatchResponse (the parent)? Makes sense as a typical rpc response. :thumbs-up:

No, still on the FetchMatchSummary.  This is for if we have any matches to return but something fails within OM core.  Eg, some but not all of the matches couldn't be added to the ignore list.


> I understand how an mmf might generate an error (like if there's something not caught and open match is going to indicate some quantity success + "run cut short by pipe closure" [or something]). But the evaluator runs across fetchMatches-requests. What's the intention of the evaluator status?

Yes, the evaluator status will be shared across all fetch match calls.  I think something I was doing subconsciously but didn't express in writing is that this is likely the first place someone will look when they're trying to get the setup working and things are broken.  So if their evaluator isn't working, this gives an obvious place in the response instead of having to start digging into logs.  The proposals and accepted helps as this as well: If you're getting no matches, it answers which piece is giving no matches.

**Comment by sawagh** [comment]
I have a set of different concerns with having 3 statuses. 

**Developer Experience** 
We would now have 3 different error status (mmf, evaluator, system) along with the API itself throwing an error. From the perspective of external developer, I think the case where you get no RPC error but mmf_status populated is easier to understand (and was the original issue description). However, I do think the evaluator_status and system_status are not obvious to me. I would be ok with some error there surfacing as just an RPC error on the API itself.

**Open Match Behavior**
For Evaluator / system status, when do we choose to fail the API v/s just populate the status? Say if we were failing to talk to redis (or say the evaluator - or mmf for that matter), I would (as an end user) expect that as an RPC failure for FetchMatches than the API succeeding with a status indicating the failure. If we try to make this a generic concept across mmf / evaluator / system - I think it is difficult to express what errors are ignorable and what are not. 

This is something we would want to get more community / TSC feedback on - however, given how close we are to v0.9 release, I think we can implement a scoped solution in v0.9 and scout for feedback on whether we expand it:

**v0.9 Suggestion**
We should have a scoped solution for the original problem - by having a mmf status, proposals generated in there. We can get feedback on that and brainstorm early in v0.10 as to whether we need more status in there. 

message FetchSummary {
  google.rpc.Status mmf_status = 1;
  int64 proposals_generated = 2;
}

User Expectation: If the MMF failed (or was cancelled), you will get that information along with additional information on number of proposals generated in a FetchMatchSummary. You can choose to ignore this and simply process the partial results.

**Comment by calebatwd** [comment]
I second @sawagh concerns with having too many status and not enough specific scenarios they support. The mmf status is critical to director development. A system status/evaluator status aren't the concerns of the consumer of the fetchMatches endpoint (a director). They're primarily the concerns of the infrastructure operations (the evaluator is misconfigured or open match is in an error'd state).

For clarification, the plan here is still to support a general standard rpc-level error... which is to say... "the open match backend was unable to complete your requests because of [insert reason, eg. invalid request, etc"

**Comment by sawagh** [comment]
_For clarification, the plan here is still to support a general standard rpc-level error... which is to say... "the open match backend was unable to complete your requests because of [insert reason, eg. invalid request, etc"?_

I think that is critical. The pattern of - something went wrong - the API failed and here is why - using the standard rpc-level error is something we should continue to do.

**Comment by Laremere** [comment]
After discussion with @sawagh there's not a clear API change that I think could be made without significantly delaying 0.9.  To quote the original motivation section:

> An alternative is to make the FetchMatches call only operate on a single profile (removing the repeated), though that still has some of the streaming problems. Though they're less severe as they're specifically limited to the erroring mmf.

This is already done.  So the main bulk of the issue has been resolved, and we're left discussing the impacts, severity, and complexity tradeoffs related to those streaming problems.

Moving this issue to the 0.10 timeframe, where I will gather a new round of info on what impact this problem still has, and present a new proposal (with overview, motiviation, impact, etc.) for review.  One important piece I see as a takeaway from this effort is that an example of calling code, eg the demo, would be very useful in showing how clients will use the API.  There's some complexity which wasn't obvious to me when writing the most recent proposal which became immediately obvious once I wrote the new code for the first match demo.

**Comment by sawagh** [comment]
@calebatwd @Laremere 

Let us discuss this in the next sync and close on the approach for v0.10

**Comment by sawagh** [comment]
Removing the assignee as in its current state, we cannot make progress. Once we have figured what to do here, we can re-assign based on bandwidth.

**Comment by sawagh** [comment]
Will write up a proposal for this one.

**Comment by sawagh** [comment]
This issue covers a lot of aspects - some of which have already been implemented. To cover the specific aspect that we want to discuss, I have created https://github.com/googleforgames/open-match/issues/1154 and have referenced this issue there. Closing this issue to avoid confusion.
